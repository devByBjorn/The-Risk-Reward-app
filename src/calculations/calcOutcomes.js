const getOutcome = (trades, outcome) =>
  trades.filter((trade) => trade.outcome === outcome)

const getOutcomeCount = (trades, outcome) =>
  getOutcome(trades, outcome).length

const getHitRatio = (trades) =>
  ((getOutcomeCount(trades, 'win') / trades.length) * 100).toFixed(2)

export { getOutcome, getOutcomeCount, getHitRatio }