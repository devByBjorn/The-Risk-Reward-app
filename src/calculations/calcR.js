const getTotalR = (trades) =>
  trades.reduce((total, trade) => total + trade.rewardToRisk, 0)

const getAvarageR = (trades) =>
  (getTotalR(trades) / trades.length).toFixed(2)

const sortByR = (trades) =>
  trades.sort((a, b) => a.rewardToRisk > b.rewardToRisk ? 1 : -1)


const getLowestR = (trades) =>
  trades.length && sortByR(trades)[0].rewardToRisk

const getHighestR = (trades) =>
  trades.length && sortByR(trades)[trades.length - 1].rewardToRisk


export { getAvarageR, getTotalR, getLowestR, getHighestR, sortByR }

