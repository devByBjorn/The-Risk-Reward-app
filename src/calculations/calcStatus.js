export const filterByStatus = (trades, status) =>
  trades.filter((trade) => trade.status === status)