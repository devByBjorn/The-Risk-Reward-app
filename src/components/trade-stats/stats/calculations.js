// R 
export const getTotalR = (trades) =>
  trades.reduce((total, trade) => total + trade.rewardToRisk, 0)

export const getAvarageR = (trades) =>
  getTotalR(trades) / trades.length

const sortByR = (trades) =>
  trades.sort((a, b) => a.rewardToRisk > b.rewardToRisk ? 1 : -1)

export const getLowestR = (trades) =>
  sortByR(trades)[0].rewardToRisk

export const getHighestR = (trades) =>
  sortByR(trades)[trades.length - 1].rewardToRisk


// SET UPS
export const getSetupsArray = (trades) =>
  trades.map((trade) => trade.setup)

export const outcomePerSetup = (trades, outcome) => {
  const tradesByOutcom = getOutcome(trades, outcome)
  const outcomeSetups = tradesByOutcom.map((trade) => trade.setup)
  const outcomeCountList = getObjectCountList(outcomeSetups)
  const mostCountedSetup = getMostOcurredKey(outcomeCountList)
  const arrOfMostCountedSetup = trades.filter((trade) =>
    trade.outcome === outcome
    && trade.setup.toLowerCase() === mostCountedSetup.toLowerCase())

  return {
    mostCountedSetup,
    arrOfMostCountedSetup
  }
}

// OUTCOMES
export const getOutcome = (trades, outcome) =>
  trades.filter((trade) => trade.outcome === outcome)

export const getOutcomeCount = (trades, outcome) =>
  getOutcome(trades, outcome).length

export const getHitRatio = (trades) =>
  (getOutcomeCount(trades, 'win') / trades.length) * 100


// COUNT LISTS
export const getObjectCountList = (array) =>
  array.length > 0
    ? array.reduce((acc, item) => {
      const count = acc[item] || 0

      return {
        ...acc,
        [item]: count + 1
      }

    }, {})
    : false

export const getMostOcurredKey = (list) =>
  !list
    ? false
    : Object.keys(list).
      reduce((a, b) => list[a] > list[b] ? a : b)

export const getLeastOcurredKey = (list) =>
  !list
    ? false
    : Object.keys(list).
      reduce((a, b) => list[a] < list[b] ? a : b)

export const objListToArr = (listAllSetups) => {
  const arr = []
  for (let [key, value] of Object.entries(listAllSetups)) {
    arr.push([key, value].join(' : '));
  }
  return arr
}