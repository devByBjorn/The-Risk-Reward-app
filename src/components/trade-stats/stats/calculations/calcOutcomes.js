const getOutcome = (trades, outcome) =>
  trades.filter((trade) => trade.outcome === outcome)

const getOutcomeCount = (trades, outcome) =>
  getOutcome(trades, outcome).length

const getHitRatio = (trades) =>
  (getOutcomeCount(trades, 'win') / trades.length) * 100

export { getOutcome, getOutcomeCount, getHitRatio }