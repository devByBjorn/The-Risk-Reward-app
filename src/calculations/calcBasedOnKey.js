import { getOutcome } from './calcOutcomes'
import { getObjectCountList, getMostOcurredKey } from './objLists'

// SET UPS
const getKeysArr = (trades, key) =>
  trades.map((trade) => trade[key])

const outcomePerSetup = (trades, outcome) => {
  const tradesByOutcom = getOutcome(trades, outcome)
  const outcomeSetups = tradesByOutcom.map((trade) => trade.setup)
  const outcomeCountList = getObjectCountList(outcomeSetups)
  const mostCountedSetup = getMostOcurredKey(outcomeCountList)
  const arrOfMostCountedSetup = trades.filter((trade) =>
    trade.outcome === outcome
    && trade.setup.toLowerCase() === mostCountedSetup.toLowerCase())

  return {
    mostCountedSetup,
    arrOfMostCountedSetup
  }
}

function getArrOfArrsTrades(trades, ArrKeys, key) {
  const trimDuplicates = [...new Set(ArrKeys)]
  const arrayOfArrays = []

  for (let i = 0; i < trimDuplicates.length; i++) {
    arrayOfArrays.push(trades.filter((trade) =>
      trade[key] === trimDuplicates[i] && trade))
  }
  return arrayOfArrays
}

export { getKeysArr, outcomePerSetup, getArrOfArrsTrades }