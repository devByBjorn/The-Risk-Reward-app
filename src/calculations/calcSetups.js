import { getOutcome } from './calcOutcomes'
import { getObjectCountList, getMostOcurredKey } from './objLists'

// SET UPS
const getSetupsArray = (trades) =>
  trades.map((trade) => trade.setup)

const outcomePerSetup = (trades, outcome) => {
  const tradesByOutcom = getOutcome(trades, outcome)
  const outcomeSetups = tradesByOutcom.map((trade) => trade.setup)
  const outcomeCountList = getObjectCountList(outcomeSetups)
  const mostCountedSetup = getMostOcurredKey(outcomeCountList)
  const arrOfMostCountedSetup = trades.filter((trade) =>
    trade.outcome === outcome
    && trade.setup.toLowerCase() === mostCountedSetup.toLowerCase())

  return {
    mostCountedSetup,
    arrOfMostCountedSetup
  }
}

function getAverageRPerSetup(trades, allSetups) {
  const noDuplicateSetupName = [...new Set(allSetups)]
  const arrayOfArrays = []

  for (let i = 0; i < noDuplicateSetupName.length; i++) {
    arrayOfArrays.push(trades.filter((trade) =>
      trade.setup === noDuplicateSetupName[i] && trade))
  }
  return arrayOfArrays
}

export { getSetupsArray, outcomePerSetup, getAverageRPerSetup }