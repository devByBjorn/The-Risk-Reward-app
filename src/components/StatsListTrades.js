import React from 'react'
import { connect } from 'react-redux'
import { getOutcomeCount, getWinRatio } from '../calculations/calcOutcomes'
import HighlightedList from '../components_style/HighlightedList'
import ListItem from '../components_style/ListItem'
import Span from '../components_style/Span'
import SubHeading from '../components_style/SubHeadingStyled'
import { StatsListWrapper } from '../components_style/StatsListWrapperStyled'

export const StatsListTrades = ({ trades }) => {
  const totalTrades = trades.length
  const winRatio = getWinRatio(trades)
  const losses = getOutcomeCount(trades, 'loss')
  const wins = getOutcomeCount(trades, 'win')
  const scratched = getOutcomeCount(trades, 'scratch')

  const listItems = [totalTrades, wins, losses, scratched, `${winRatio}%`]
  const itemsHeadings = ['Total', 'Wins', 'Losses', 'Scratched', 'WinRatio']

  return (
    <StatsListWrapper flexDirection="column">
      <SubHeading>Trades</SubHeading>
      <HighlightedList>
        {listItems.map((item, i) => (
          <ListItem key={item}>
            <Span>{itemsHeadings[i]}</Span>
            {trades.length ? item : '-'}
          </ListItem>
        ))}
      </HighlightedList>
    </StatsListWrapper>
  )
}

const mapStateToProps = (state) => ({
  trades: state.trades.filter((trade) => trade.status === 'closed')
})

export default connect(mapStateToProps)(StatsListTrades)