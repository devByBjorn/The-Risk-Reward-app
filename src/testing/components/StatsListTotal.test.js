import React from 'react'
import { shallow } from 'enzyme'
import { StatsListTrades } from '../../components/StatsListTrades'
import { trades } from '../fixtures/trades'
import { getOutcomeCount, getWinRatio } from '../../calculations/calcOutcomes'

test('should render StatsListTrades component', () => {
  const wrapper = shallow(<StatsListTrades trades={trades} />)
  expect(wrapper).toMatchSnapshot()
})

test('should return count of winning trades', () => {
  expect(getOutcomeCount(trades, 'win')).toBe(2)
})

test('should return win ratio of all trades', () => {
  expect(getWinRatio(trades)).toBe(66.67)
})