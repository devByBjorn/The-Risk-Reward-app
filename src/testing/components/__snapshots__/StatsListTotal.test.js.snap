// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render StatsListTrades component 1`] = `
<StatsListWrapperStyled__StatsListWrapper
  flexDirection="column"
>
  <SubHeadingStyled__SubHeading>
    Trades
  </SubHeadingStyled__SubHeading>
  <HighlightedList>
    <ListItem
      key="3"
    >
      <Span>
        Total
      </Span>
      3
    </ListItem>
    <ListItem
      key="2"
    >
      <Span>
        Wins
      </Span>
      2
    </ListItem>
    <ListItem
      key="1"
    >
      <Span>
        Losses
      </Span>
      1
    </ListItem>
    <ListItem
      key="0"
    >
      <Span>
        Scratched
      </Span>
      0
    </ListItem>
    <ListItem
      key="66.67%"
    >
      <Span>
        WinRatio
      </Span>
      66.67%
    </ListItem>
  </HighlightedList>
</StatsListWrapperStyled__StatsListWrapper>
`;
